nixos/test-driver: use vhost-device-vsock for SSH backdoor

Ma27 opened the pull request:
__A few notes for reviewers:__

* The last commit is doing the relevant stuff. The commits before are essentially preparations and cleanups.
* We need systemd 258 (#427968) for this to work out of the box with a global ssh config and openssh 10.2p1 for the login to be fast (#450658).

Since I'm on systemd 257, I used the following ssh config:

```
Host vsock-mux/*
        ProxyCommand /run/current-system/sw/lib/systemd/systemd-ssh-proxy %h %p
        ProxyUseFdpass yes
        CheckHostIP no

        # Disable all kinds of host identity checks, since these addresses are generally ephemeral.
        StrictHostKeyChecking no
        UserKnownHostsFile /dev/null
```

Because of that and with this being an invasive last-minute change, I'd prefer to target 26.05 instead of 25.11 here.

Todo:

* [ ] run more VM tests before merging.

---

`vhost-device-vsock`[1] is a custom implementation of AF_VSOCK, but the
application on the host-side uses a UNIX domain-socket. This gives us
the following nice properties:

* We don't need to do `--arg sandbox-paths /dev/vhost-vsock` anymore for
  debugging builds within the sandbox. That means, untrusted users can
  also debug these kinds of tests now.

* This prevents CID conflicts on the host-side, i.e. there's no need for
  using `sshBackdoor.vsockOffset` for tests anymore.

A big shout-out goes to Allison Karlitskaya, the developer of test.thing[2]
who talked about this approach to do AF_VSOCK on All Systems Go 2025.

This patch requires systemd 258[3] because this contains `vhost-mux` in
its SSH config which is needed to connect to the VMs from now on.

To not blow up the patches even more, this only uses AF_VSOCK for the
debugger. A potential follow-up for the future would be a removal of the
current `backdoor.service` and replace it entirely by this
functionality.

[1] https://github.com/rust-vmm/vhost-device/blob/main/vhost-device-vsock/README.md
[2] https://codeberg.org/lis/test.thing
[3] #427968

---

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 committed:
vhost-device-vsock: init at 0.3.0

Ma27 committed:
nixos/testing: allow setting test-wide warnings and assertions

This allows us to e.g. use `mkRemovedOptionModule` which will come in
handy in the upcoming commits.

Ma27 committed:
nixos/qemu-vm: add `enableSharedMemory` option

This option configures a memfd backend for the VM's memory with the size
of `virtualisation.memorySize` and uses that as default memory backend.

This is required for e.g. vhost-device-vsock.

The motivation for making this an option is that I decided to enable
this by default for all NixOS tests to avoid changing essential QEMU
options based on whether or not debugging is enabled. However, there
should be an easy way to turn it off which is what this option provides.

tfc commented on nixos/lib/test-driver/src/test_driver/machine/__init__.py:
```diff
@@ -147,6 +150,8 @@ def cmd(
         if not allow_reboot:
             qemu_opts += " -no-reboot"
 
+        qemu_opts += vhost_vsock.qemu_args
```
The `StartCommand` class is already extremely educated about qemu specific things.
The `VHostVsock` class encapsulates very qemu specific concerns.

The sum of LOC that we get with this abstraction is seemingly higher than if we would take a `vsock_id: int | None` as a parameter here and then did something like

```python
if vsock_id is not None:
  qemu_opts += "--vsock-foo-bar-{vsock_id}"
```

At the same time this additional abstration doesn't buy us much and qualifies as [shallow module or "classitis"](https://galowicz.de/2022/12/05/book-review-a-philosophy-of-software-design/).

Especially looking at the future where we might have different backends for the Machine class, it might not be helpful to scatter qemu concerns more.

I suggest to remove the `VHostVsock` class and use the simpler approach. (Respectfully, assuming i did not miss any concerns while i was reading the code)

Ma27 replied:
I changed the approach a little bit. As documented in the last commit, there are two new classes:

* `VsockPair` which contains three properties. The goal here is to not use triples for a host-path/guest-path/guest-cid since having two times `Path` in it has the potential of mixing up things.
* `VsockDeviceVsock` abstracts away the creation of the socket pairs and the start of the process itself. I think that's not necessarily a detail, the driver should take care about (similar to e.g. starting up VLANs). This isn't its own context-manager anymore, it uses the context-manager rom the driver, like everything else which seems much cleaner.
* The QEMU stuff is part of StartCommand, as suggested. That wayh, we only have to pass `guest_path` to the machine (and start command), nothing else which makes another part of the abstraction go away.
* I noticed it's trivial now to write a test to make sure the ssh backdoor works, so we have a test now :)

Ma27 committed:
nixos/test-driver: push allocation into the context-manager

The context manager's purpose is to allocate its resources in `__enter__` and
release them again in `__exit__`.

Right now, the approach is merely a hack since we allocate everything in
the constructor, but use the context-manager protocol as way to reliably
terminate everything.

This isn't a functional change, but merely a correctness change by using
the methods the way they were intended.

Ma27 committed:
nixos/test-driver: use vhost-device-vsock for SSH backdoor

`vhost-device-vsock`[1] is a custom implementation of AF_VSOCK, but the
application on the host-side uses a UNIX domain-socket. This gives us
the following nice properties:

* We don't need to do `--arg sandbox-paths /dev/vhost-vsock` anymore for
  debugging builds within the sandbox. That means, untrusted users can
  also debug these kinds of tests now.

* This prevents CID conflicts on the host-side, i.e. there's no need for
  using `sshBackdoor.vsockOffset` for tests anymore.

A big shout-out goes to Allison Karlitskaya, the developer of test.thing[2]
who talked about this approach to do AF_VSOCK on All Systems Go 2025.

This patch requires systemd 258[3] because this contains `vhost-mux` in
its SSH config which is needed to connect to the VMs from now on.

To not blow up the patches even more, this only uses AF_VSOCK for the
debugger. A potential follow-up for the future would be a removal of the
current `backdoor.service` and replace it entirely by this
functionality.

The internal implementation tries to be consistent with how VLANs and
machines are handled, i.e. the processes are started when the Driver's
context is entered and cleaned up in __exit__().

I decided to push the process management and creation of sockets for
vhost-device-vsock into its own class, that's an implementation detail
and not a concern for the test-driver. In fact, `vhost-device-vsock` is
something we can drop once QEMU implements native support for using
AF_UNIX on the host-side[4]. `VsockPair` is its own class since
returning e.g. a triple of `(Path, Path, Int)` would be ambiguous in
what is the guest and what the host path (and frankly, I found it hard
to distinguish the two when reading the docs of `vhost-device-vsock`
initially).

Finally, now that we can do the SSH backdoor without adding additional
devices to the sandbox, I figured, it's time to write a test-case for
it.

[1] https://github.com/rust-vmm/vhost-device/blob/main/vhost-device-vsock/README.md
[2] https://codeberg.org/lis/test.thing
[3] https://github.com/NixOS/nixpkgs/pull/427968
[4] https://gitlab.com/qemu-project/qemu/-/issues/2095

tfc commented on nixos/lib/test-driver/src/test_driver/driver.py:
```diff
@@ -272,18 +341,28 @@ def subtest(name: str) -> Iterator[None]:
+                f"{Style.BRIGHT}Note:{Style.RESET_ALL} this requires {Style.BRIGHT}systemd-ssh-proxy(1){Style.RESET_ALL} to be enabled (default on NixOS 25.05 and newer)."
+            )
+            longest_name = len(
+                max((machine.name for machine in self.machines), key=len)
```
this is not important as we're not operating on much data, but aren't we calculating string lengths here twice?

i'd suggest `max(len(m.name) for m in self.machines))`

Ma27 replied:
Correct. I feel like that's not a sufficiently big deal given this code-path is only reached when turning on debugging features, the code seems easier to follow (to me at least) with the `key=len` thing rather than precomputing all the lengths up-front and re-use them twice and finally, we can't have large data in there anyways since other parts of the test-driver are not designed for that: https://github.com/NixOS/nixpkgs/issues/488423

So my preference would be to just accept that.

tfc commented on nixos/tests/nixos-test-driver/ssh-backdoor.nix:
```diff
@@ -0,0 +1,41 @@
+            "-o",
+            "User=root",
+            "-F",
+            # FIXME not needed anymore with systemd 258
```
it sounds like we could remove this now, is that true?

Ma27 replied:
Thanks for the reminder. Will do throughout the day and let you know :)

tfc commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -515,16 +515,16 @@ with SSH over vsock:
-which can be done with e.g.
+This will halt the test execution on a test-failure and print instructions
+on how to enter the sandbox shell of the VM test. Inside, one can log into
+e.g. `machine` with
```
not a big thing but it seems like the characteristic that a test is halted after error is already mentioned above. this paragraph here seems like it should only be about connecting. and then the command before this paragraph and the one after this paragraph seem somehow redundant and/or the command above seems to be old?

Ma27 replied:
I wonder if I just managed to mess up the rebase and that's why duplicated information is in here. Fixing.

tfc commented on nixos/lib/test-driver/src/test_driver/driver.py:
```diff
@@ -63,19 +64,64 @@ def pythonize_name(name: str) -> str:
+        }
+
+        self.vhost_proc = subprocess.Popen(
+            (
```
just a minor nitpick but we're using lists in all other Popen calls in the code.

Ma27 commented:
@tfc addressed your notes now :)
